### PR TITLE
Move GpuBlasLtAdaptor out

### PR DESCRIPTION
**Pull request by alekstheod**

### CODE REVIEW

File: third_party/xla/xla/stream_executor/gpu/gpu_blas_lt_adaptor.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+#include <cstdint>
+#include <utility>
+
+#include "xla/stream_executor/blas.h"
+#include "xla/stream_executor/gpu/gpu_blas_lt_gemm_runner.h"
+
+namespace stream_executor::gpu {
+
+template <typename TBlasSupport>
+struct GpuBlasLtAdaptor final : TBlasSupport {
```
Comment (i-chaochen) [inline third_party/xla/xla/stream_executor/gpu/gpu_blas_lt_adaptor.h:56]: not sure we really need `final` .. @pemeliya do you think is it ok?

File: third_party/xla/xla/stream_executor/gpu/gpu_blas_lt_adaptor.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    } else {
+      return TBlasSupport::DoBlasGemmBatched(
+          stream, transa, transb, m, n, k, alpha, a, lda, b, ldb, beta, c, ldc,
+          batch_count, numeric_options, scratch_allocator, context);
+    }
+  }
+
+ private:
+  bool CheckStatus(absl::Status status) {
+    return status.code() == absl::StatusCode::kOk;
```
Comment (i-chaochen) [inline third_party/xla/xla/stream_executor/gpu/gpu_blas_lt_adaptor.h]: like this with LOG(ERROR)? https://github.com/ROCm/tensorflow-upstream/blob/r218-hipblaslt-nonxla-1126/third_party/xla/xla/stream_executor/stream_common.cc#L155-L158

File: third_party/xla/xla/stream_executor/gpu/gpu_blas_lt_adaptor.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  bool IsGpuBlasLtEnabled() { return false; }
```
Comment (i-chaochen) [inline third_party/xla/xla/stream_executor/gpu/gpu_blas_lt_adaptor.h]: I think it's just temporary one?

File: third_party/xla/xla/stream_executor/gpu/gpu_blas_lt_adaptor.h [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+#include <cstdint>
+#include <utility>
+
+#include "xla/stream_executor/blas.h"
+#include "xla/stream_executor/gpu/gpu_blas_lt_gemm_runner.h"
+
+namespace stream_executor::gpu {
+
+template <typename TBlasSupport>
```
Comment (i-chaochen) [inline third_party/xla/xla/stream_executor/gpu/gpu_blas_lt_adaptor.h:55]: IIUC, to use this CRTP-based adaptor [here in matmul_op_impl.h](https://github.com/ROCm/tensorflow-upstream/blob/r218-hipblaslt-nonxla-1126/tensorflow/core/kernels/matmul_op_impl.h#L729-L730) we should also need `BaseBlasSupport` and calling like this?
```
  // Specify the template type explicitly
  GpuBlasLtAdaptor<BaseBlasSupport> adaptor;

  // Call the method
  bool success = adaptor.DoBlasGemmBatched();
```

**Review by i-chaochen — COMMENTED** [review]
Thanks a lot!

File: third_party/xla/xla/stream_executor/gpu/gpu_blas_lt_adaptor.h (hunk 4 above)
Comment (alekstheod) [reply third_party/xla/xla/stream_executor/gpu/gpu_blas_lt_adaptor.h:55]: Real adaptor would be better than just overriding the concrete function. However there are so many overloads that need to be duplicated just to propagate the call further. I thought about generating them through the template helper recursively but it makes everything complicated. So the simplest solution is to just inherit.

File: third_party/xla/xla/stream_executor/gpu/gpu_blas_lt_adaptor.h (hunk 3 above)
Comment (alekstheod) [reply third_party/xla/xla/stream_executor/gpu/gpu_blas_lt_adaptor.h]: Yes it is.

File: third_party/xla/xla/stream_executor/gpu/gpu_blas_lt_adaptor.h (hunk 4 above)
Comment (alekstheod) [reply third_party/xla/xla/stream_executor/gpu/gpu_blas_lt_adaptor.h:55]: Also is this only valid for ROCM? If so we don't even need to make it template. We could directly inherit from ROCMBlas

File: tensorflow/core/kernels/matmul_op_impl.h [hunk 5]
```diff
@@ -725,8 +725,12 @@ struct LaunchBatchMatMul<GPUDevice, Scalar> {
 
         BlasScratchAllocator scratch_allocator(context, max_scratch_size);
         auto blas = stream->parent()->AsBlas();
+        
         OP_REQUIRES(context, blas != nullptr,
                     absl::InternalError("No blas support for stream"));
+        
+        // TODO: calling hipblaslt here to replace blas->DoBlasGemmBatched
```
Comment (i-chaochen) [inline tensorflow/core/kernels/matmul_op_impl.h]: I guess this is copied from https://github.com/ROCm/tensorflow-upstream/commit/fbef0b5a170e204c06eb035911f043d7abbfd5f7 we can remove this now.

File: third_party/xla/xla/stream_executor/blas.h [hunk 6]
```diff
@@ -717,6 +717,32 @@ class BlasSupport {
   void operator=(const BlasSupport &) = delete;
 };
 
+// struct GpuBlasLtAdaptor : public BlasSupport {
```
Comment (i-chaochen) [inline third_party/xla/xla/stream_executor/blas.h]: ditto

**Comment by i-chaochen** [comment]
Hi @alekstheod I cannot see gpu-pycpp results now, is this branch PASSED all unit tests now? if CI is instable recently could you test single-gpu and multi-gpus on your locally and let's merge this PR.

File: third_party/xla/xla/stream_executor/blas.h (hunk 6 above)
Comment (alekstheod) [reply third_party/xla/xla/stream_executor/blas.h]: Remoted

File: tensorflow/core/kernels/matmul_op_impl.h (hunk 5 above)
Comment (alekstheod) [reply tensorflow/core/kernels/matmul_op_impl.h]: Removed

**Comment by alekstheod** [comment]
> Hi @alekstheod I cannot see gpu-pycpp results now, is this branch PASSED all unit tests now? if CI is instable recently could you test single-gpu and multi-gpus on your locally and let's merge this PR.

I triggered the CI again. If it is red due to CI I will test locally.

**Comment by alekstheod** [comment]
gpu-pycpp is green. Seems that CI is not stable as in a previous run we had gpu-non-pip-multi green and the rest was red?

**Comment by alekstheod** [comment]
Tried that locally and the test is green . Also tried to call std::abort inside the adaptor. It seems adaptor is not called during that test exection. I reverted irrelevant version change. Maybe that version does affect the test.

**Comment by i-chaochen** [comment]
Thanks! if gpu-non-pip-multi is no problem we can merge this PR

**Review by i-chaochen — APPROVED** [review]

**Pull request merged**